File: herencia-dto/HerenciaDto-Java-IvanRC/src/main/java/storage/personas/PersonaStorageServiceCsv.java
package storage.personas;

import config.ConfigApp;
import dto.PersonaDto;
import mapper.PersonaMapper;
import models.Persona;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class PersonaStorageServiceCsv implements PersonasStorageService{

    ConfigApp configApp = ConfigApp.getInstance();

    File file = new File(configApp.getAPP_DATA()+File.separator+"personas.csv");

    public PersonaStorageServiceCsv() throws IOException {}

    @Override
    public void saveAll(List<Persona> entities) {
        try(BufferedWriter br = new BufferedWriter(new FileWriter(file))) {
            br.write("nombre;edad;modulo;tipo\n");
            entities.stream().map(p -> PersonaMapper.toDto(p)).forEach(persona ->
                    {
                        try {
                            br.append(
                                    persona.getNombre()+";"+persona.getEdad()+";"+persona.getModulo()+";"+persona.getTipo()+"\n"
                            );
                        } catch (IOException e) {
                            throw new RuntimeException(e);
                        }
                    }
            );
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public List<Persona> loadAll() throws FileNotFoundException {
        List<Persona> ingredientes = new ArrayList<>();
        if(!file.exists()) return ingredientes;
        try(BufferedReader br = new BufferedReader(new FileReader(file))){
            String line = br.readLine();
            while ((line = br.readLine()) != null) {
                String[] campos = line.split(";");
                ingredientes.add(PersonaMapper.toPersona(new PersonaDto(campos[0], campos[1], campos[2], campos[3])));
            }
        } catch (FileNotFoundException e) {
            throw new RuntimeException(e);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return ingredientes;
    }
}

File: herencia-dto/HerenciaDto-Kotlin-IvanRC/src/main/kotlin/storage/persona/PersonaStorageServiceCsv.kt
package storage.persona

import config.ConfigApp
import dto.PersonaDto
import mapper.toPersona
import mapper.toPersonasDto
import models.Persona
import mu.KotlinLogging
import java.io.File

class PersonaStorageServiceCsv: PersonaStorageService {

    private val configApp = ConfigApp
    private val file = File(configApp.APP_DATA+ File.separator+"personas.csv")

    private val logger = KotlinLogging.logger {  }

    override fun saveAll(entities: List<Persona>) {
        logger.debug { "Se guardan todas las personas en el fichero CSV" }
        file.writeText("nombre;edad;modulo;tipo\n")
        entities.toPersonasDto().personasDto.forEach {
            file.appendText(
                "${it.nombre};${it.edad ?: ""};${it.modulo ?: ""};${it.tipo}\n"
            )
        }
    }

    override fun loadAll(): List<Persona> {
        logger.debug { "Se cargan todas las personas del fichero CSV" }
        return file.readLines().drop(1)
            .map { it.split(";") }
            .map { it.map { it.trim() } }
            .map {campos ->
                PersonaDto(
                    nombre = campos[0],
                    edad = if(campos[1] != "") campos[1] else null,
                    modulo = if(campos[2] != "") campos[2] else null,
                    tipo = campos[3]
                )
            }
            .map { it.toPersona() }
    }
}